### PR TITLE
Migrate and adapt Windows wazuh agent packages subsystem

### PR DESCRIPTION
**Pull request by jotacarma90**
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/22080|

## Description
PR to add workflows and script changes needed to Windows package creation

### CODE REVIEW

**Comment by jotacarma90** [comment]
# Update 03/04/2024
- Multiple workflow changes to correctly handle the package name.
- Fixed problems uploading and downloading artifact with compiled code.-
- Added smoke test to verify that the package has been installed correctly (fixing some pending issues).
- Modifications in the msi generation script to set the package name by parameter.
- Pending to include changes for package signing: https://github.com/wazuh/wazuh/issues/21755#issuecomment-2034819851

File: .github/workflows/packages-build-windows-agent.yml [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        uses: actions/download-artifact@v4
+        with:
+          name: windowsPKG
+          path: C:\
+
+      - name: Unzip base folder
+        run: |
+          7z x C:\${{ needs.build-agent-windows-base.outputs.PKG_NAME }}.zip -oC:\win-agent-base
+
+      - name: Create .msi package
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/workflows/packages-build-windows-agent.yml]: done

File: .github/workflows/packages-build-windows-agent.yml [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        description: |
+          Package revision (name and metadata)
+          Default is '0'.
+        required: false
+        default: '0'
+      is_stage:
+        description: |
+          Should set development/production nomenclature
+          True if WPK name should have production format.
+          False if WPK name should have developer format.
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/workflows/packages-build-windows-agent.yml]: done

File: .github/workflows/packages-build-windows-agent.yml [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+          Default is '1'.
+        options:
+          - 0
+          - 1
+          - 2
+        required: false
+        default: '1'
+      ca_name:
+        description: |
+          CA name to be used to verify the trust of the agent."
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/workflows/packages-build-windows-agent.yml]: done

File: .github/workflows/packages-build-windows-agent.yml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+          cat C:\win-agent-base\installer.log
+          if($SEL -eq $null) {
+            exit 1
+          }
+
+      - name: Upload msi to s3
+        run: |
+          aws s3 cp C:\win-agent-base\${{ needs.build-agent-windows-base.outputs.PKG_NAME }}.msi s3://packages-dev.internal.wazuh.com/development/wazuh/4.x/main/packages/
+
+      - name: Upload checksum to s3
```
[the remark this answers is not in the document]
Comment (jotacarma90) [reply .github/workflows/packages-build-windows-agent.yml]: done

File: .github/workflows/packages-build-windows-agent.yml [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        default: '0'
+      is_stage:
+        description: |
+          Should set development/production nomenclature
+          True if package name should have production format.
+          False if package name should have developer format.
+          Default is 'false'.
+        type: boolean
+        required: false
+        default: false
```
Comment (lchico) [inline .github/workflows/packages-build-windows-agent.yml:27]: These lines can be safely removed as they are already set to false by default. Leaving a comment explaining this (as you've done) is sufficient.

File: .github/workflows/packages-build-windows-agent.yml [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+          if [ "${{ inputs.docker_image_tag }}" == "auto" ]; then echo "TAG=${{ env.VERSION }}" >> $GITHUB_ENV;
+          elif [ "${{ inputs.docker_image_tag }}" == "developer" ]; then echo "TAG=${{ github.ref_name }}" >> $GITHUB_ENV;
+          else echo "TAG=${{ inputs.docker_image_tag }}" >> $GITHUB_ENV; fi
+
+      - name: Set PKG_NAME
+        id: package_name_set_step
+        run: |
+          COMMIT_HASH="$(git rev-parse --short HEAD)"
+          if [ "${{ inputs.is_stage }}" == "true" ]; then
+            echo "PKG_NAME=wazuh-agent-${{ env.VERSION }}-${{ inputs.revision }}" | tee -a $GITHUB_OUTPUT $GITHUB_ENV;
```
Comment (lchico) [inline .github/workflows/packages-build-windows-agent.yml:92]: Nice use of tee!

File: packages/windows/generate_compiled_windows_agent.sh [hunk 7]
```diff
@@ -105,15 +95,53 @@ main() {
...
 
-    if [[ "$BUILD" != "no" ]]; then
-        generate_compiled_win_agent || exit 1
+    if [ ! -d "${OUTDIR}" ]; then
+        echo "Creating building directory at ${OUTDIR}"
+        mkdir -p ${OUTDIR}
+    fi
+
+    if [ -z "${CUSTOM_CODE_VOL}" ]; then
+        cd ../..
```
Comment (lchico) [inline packages/windows/generate_compiled_windows_agent.sh:129]: We could add a TODO: This needs some work. Consider validating the path to the Wazuh directory (e.g., `*/wazuh*/packages/windows`)

**Review by lchico — APPROVED** [review]
Looks great! To improve validation, let's add checks for the CUSTOM_CODE_VOL path.

**Review by MarcelKemp — APPROVED** [review]

**Comment by jotacarma90** [comment]
## Builds successfully with different params

🟢 https://github.com/wazuh/wazuh/actions/runs/8570640653
🟢 https://github.com/wazuh/wazuh/actions/runs/8570725814

**Pull request merged**